product-service/src/functions/catalogBatchProcess.ts

```diff
@@ -0,0 +1,59 @@
+import { ProductNotificationService } from '@libs/services/notificationService';
+import { ProductService } from '@libs/services/productsService';
+import Ajv from 'ajv';
+import { SQSEvent, SQSHandler } from 'aws-lambda';
+import { ProductRequest, ProductRequestSchema } from '../dto/product';
+
+const validateProduct = (new Ajv()).compile(ProductRequestSchema);
+
+const catalogBatchProcess = (productService: ProductService, notificationService: ProductNotificationService): SQSHandler => async (event: SQSEvent)=> {
+  try {
+    console.log(`Incoming event: ${ JSON.stringify(event) }`);
+
+    const products = event.Records.reduce((acc, record) => {
+      let productRequest: unknown;
+
+      try {
+        productRequest = JSON.parse(record.body) as ProductRequest;
+      } catch(err) {
+        productRequest = {};
+      }
+
+      if (!validateProduct(productRequest)) {
+        return acc;
+      }
+
+      return [...acc, productRequest as ProductRequest];
+    }, [] as ProductRequest[]);
+
+    if (!products.length) {
+      console.log('There are no valid products in the request');
+
+      return;
+    }
+
+    await Promise.all(products.map(async(product) => {
```

Comment:
question: any reason for Promise.all here? products.map should return the array of promises to launch them resolving async. In this case, we are still waiting for the await creation of a product and await sendNotification every new map iteration,  aren't we?

Reply:
suggestion: if even the list of products will be launched async with Promise.all all products lost in case of one promise is failed. Better to use Promise.allSettled().

Reply:
Hi @IgorByl, thanks for the review!  The await is needed here to guarantee the execution of `send notification` logic for products. So, for example, when this lambda function would be updated and new logic will be added after sending notifications logic we can guarantee that it will be executed after async execution of notification messages. Plus, we can catch errors in the main try / catch block for this lambda